internal/imagedestination/wrapper.go

```diff
@@ -0,0 +1,54 @@
+package imagedestination
+
+import (
+	"context"
+	"io"
+
+	"github.com/containers/image/v5/internal/types"
+	publicTypes "github.com/containers/image/v5/types"
+)
+
+// FromPublic(dest) returns an object that provides the internaltypes.ImageDestinationWithOptions API
+//
+// Eventually, we might want to expose this function, and methods of the returned object,
+// as a public API (or rather, a variant that does not include the already-superseded
+// methods of publicTypes.ImageDestination, and has added more future-proofing), and more strongly
+// deprecate direct use of publicTypes.ImageDestination.
+//
+// NOTE: The returned API MUST NOT be a public interface (it can be either just a struct
+// with public methods, or perhaps a private interface), so that we can add methods
+// without breaking any external implementors of a public interface.
+func FromPublic(dest publicTypes.ImageDestination) types.ImageDestinationWithOptions {
+	if dest2, ok := dest.(types.ImageDestinationWithOptions); ok {
+		return dest2
+	}
+	return &wrapped{ImageDestination: dest}
+}
+
+// wrapped provides the ImageDestinationWithOptions operations
+// for a destination that only implements publicTypes.ImageDestination
+type wrapped struct {
+	publicTypes.ImageDestination
+}
+
+// PutBlobWithOptions writes contents of stream and returns data representing the result.
+// inputInfo.Digest can be optionally provided if known; if provided, and stream is read to the end without error, the digest MUST match the stream contents.
+// inputInfo.Size is the expected length of stream, if known.
+// inputInfo.MediaType describes the blob format, if known.
+// WARNING: The contents of stream are being verified on the fly.  Until stream.Read() returns io.EOF, the contents of the data SHOULD NOT be available
+// to any other readers for download using the supplied digest.
+// If stream.Read() at any time, ESPECIALLY at end of input, returns an error, PutBlob MUST 1) fail, and 2) delete any data stored so far.
+func (w *wrapped) PutBlobWithOptions(ctx context.Context, stream io.Reader, inputInfo publicTypes.BlobInfo, options types.PutBlobOptions) (publicTypes.BlobInfo, error) {
```

Comment:
It would be interesting to have the internal names clean, without the `WithOptions` baggage. We can’t quite do that directly in the returned `ImageDestination` objects, and doing it here would require wrapping the `publicTypes.ImageDestination` in another non-zero-cost layer of indirection. Or, alternatively, we could have all of the internal code clean and work directly with internal objects and internal names, but always proxy external callers though the external types in separate objects (so that we can have direct access to internal `PutBlob(new)` and external callers get a proxy object with `PutBlob(old)`).

A cheap compromise would be to just name the internal methods `PutBlob2` or `PutBlobX` (with `X` reserved for all internal names), admittedly that’s also pretty ugly.